### PR TITLE
degree of separation- v1

### LINKED ISSUE #22: Implement user degree of separation

### PR DESCRIPTION
**Pull request by liyaabebe**
I've setup  the view, controller and model for displaying degree of separation. I'm not able to test it tho since I'm having trouble getting the currently logged in user. 

What we had so far was setting the current user in the model but once it is accessed, it can't be accessed again. I'll need to access the logged in user in two html pages( viewUsers.html and followSucess.html). We need to access it multiple times in the user profile page too.

Let me know if you have any solution for this!

### CODE REVIEW

**Comment by noahmank** [comment]
Regarding your problem with maintaining user state, there are two options that might solve your problem, if I understand it correctly:
1. Adding the SessionAttributes annotation for spring. Should allow you to maintain the User in the controller, but I believe it is not useful for sharing attributes across controllers (not sure if we would need to).
2. HTTPSession with setAttribute and getAttribute, haven't done too much research into how it works but I think it is a viable solution

File: src/main/java/application/controllers/UserController.java [hunk 1]
```diff
@@ -87,5 +84,27 @@ public String reviewsWritten(@PathVariable long id, Model model) {
...
+        // for testing, will be removed
+        User user1 = new User("Noah", "4321");
+        User user2 = new User("Aubin", "4321");
+        User user3 = new User("Liya", "4321");
+        users.add(user1);
+        users.add(user2);
+
+        model.addAttribute("userslist", users);
+        return "viewusers";
+    }
```
Comment (noahmank) [inline src/main/java/application/controllers/UserController.java]: Idk if this is what you were referring to when you mentioned issues accessing the current user, but this does lead to an error page when deployed.

File: src/main/java/application/models/User.java [hunk 2]
```diff
@@ -93,4 +93,28 @@ public void addFollower(User follower){
...
+                 userList2nd.add((User) userInList.getFollowingList());
+             }
+
+             if(userList2nd.contains(user)){
+                 return "2nd";
+             }
+         }
+
+        return "3rd+";
+    }
```
Comment (noahmank) [inline src/main/java/application/models/User.java]: We should probably go past just two degrees of separation (most likely up to 6 considering "6 degrees of kevin bacon").  I would also recommend writing this a little more "expandably". Could probably also use recursion, but there's not much need.  For example (assume MAX_DEGREES is declared already):

```
public int getDegreesOfSeparation(User user) {
  Set<User> directAcquaintances = new HashSet<>();
  Set<User> checked = new HashSet<>();
  directAcquaintances.addAll(this.followersList);
  directAcquaintances.addAll(this.followingList);
  for(int degrees = 1; degrees <= MAX_DEGREES; degrees++) {
    if(directAcquaintances.contains(user)) {
      return degrees;
    }
    else {
      checked.addAll(directAcquaintances);
      for(User u : new HashSet<>(directAcquaintances)) {
        directAcquaintances.addAll(u.getFollowersList);
        directAcquaintances.addAll(u.getFollowingList);
      }
      directAcquaintances.removeAll(checked);
  }
  return -1 // Signals that its more than MAX_DEGREES of separation
}
```

Probably some errors with this but for a general idea.

**Review by noahmank — CHANGES_REQUESTED** [review]
We definitely need to wait to get this pr in until after viewUsers is fixed and tests are done for degrees of separation. Just a reminder that you don't have to test degrees of separation as part of the web path, you can (and should) test that method by itself.

File: src/main/java/application/models/User.java (hunk 2 above)
Comment (liyaabebe) [reply src/main/java/application/models/User.java]: sounds good, I've updated the method!

File: src/main/java/application/controllers/UserController.java (hunk 1 above)
Comment (liyaabebe) [reply src/main/java/application/controllers/UserController.java]: will fix it once the HTTPSession pr is merged

**Comment by aubins11** [comment]
Looking good to me!:)

**Review by adelatullio — APPROVED** [review]

**Review by aubins11 — APPROVED** [review]

**Pull request merged**